### PR TITLE
add arg builder support to copyObject API

### PR DESCRIPTION
**Pull request by balamurugana**

### CODE REVIEW

File: api/src/main/java/io/minio/MinioClient.java [hunk 1]
```diff
@@ -2321,48 +2323,184 @@ public void copyObject(
...
+      }
+
+      value = map.get("x-amz-copy-source-if-unmodified-since");
+      if (value != null) {
+        builder.srcUnmodifiedSince(ZonedDateTime.parse(value, Time.HTTP_HEADER_DATE_FORMAT));
+      }
+
+      value = map.get("x-amz-metadata-directive");
+      if (value != null) {
+        builder.metadataDirective(Directive.REPLACE);
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/MinioClient.java:2361]:
```suggestion
        builder.metadataDirective(value);
```

File: api/src/main/java/io/minio/MinioClient.java [hunk 2]
```diff
@@ -2321,48 +2323,184 @@ public void copyObject(
...
+    args.validateSse(this.baseUrl);
+
+    Multimap<String, String> headers = args.genHeaders();
+
     // Source object name is optional, if empty default to object name.
-    if (srcObjectName == null) {
-      srcObjectName = objectName;
+    String srcObject = args.object();
+    if (args.srcObject() != null) {
+      srcObject = args.srcObject();
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/MinioClient.java:2450]: Can be simplified using the `Optional` class as follows:

`String srcObject = Optional.ofNullable(args.srcObject()).orElse(args.object());`

File: functional/FunctionalTest.java [hunk 3]
```diff
@@ -2299,660 +2307,368 @@ public static void threadedPutObject() throws Exception {
...
-   * copyConditions) with ETag to match.
-   */
-  public static void copyObject_test2() throws Exception {
-    if (!mintEnv) {
-      System.out.println(
-          "Test: copyObject(String bucketName, String objectName, String destBucketName,"
-              + "CopyConditions copyConditions) with Matching ETag (Negative Case)");
+    String srcObject = args.object();
+    if (args.srcObject() != null) {
+      srcObject = args.srcObject();
```
Comment (anjalshireesh) [inline functional/FunctionalTest.java:2320]: Can be simplified using `Optional` class as follows:

`String srcObject = Optional.ofNullable(args.srcObject()).orElse(args.object());`

File: api/src/main/java/io/minio/ObjectWriteArgs.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    headers.putAll(userMetadata);
+
+    if (sse != null) {
+      headers.putAll(Multimaps.forMap(sse.headers()));
+    }
+
+    String tagging =
+        tags.get().entrySet().stream()
+            .map(e -> S3Escaper.encode(e.getKey()) + "=" + S3Escaper.encode(e.getValue()))
+            .collect(Collectors.joining("&"));
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply api/src/main/java/io/minio/ObjectWriteArgs.java:76]: ATM `io.minio.S3Escaper` is private and not available to `io.minio.messages` package.

File: api/src/main/java/io/minio/ObjectWriteArgs.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    return headers;
+  }
+
+  protected void validateSse(HttpUrl url) {
+    checkSse(sse, url);
+  }
+
+  public abstract static class Builder<B extends Builder<B, A>, A extends ObjectWriteArgs>
+      extends ObjectArgs.Builder<B, A> {
+    @SuppressWarnings("unchecked") // Its safe to type cast to B as B is inherited by this class
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply api/src/main/java/io/minio/ObjectWriteArgs.java:101]: Done

File: api/src/main/java/io/minio/MinioClient.java (hunk 1 above)
Comment (balamurugana) [reply api/src/main/java/io/minio/MinioClient.java:2361]: No. `CopyConditions` sets only `"REPLACE"` string. Hence `Directive.REPLACE`

File: api/src/main/java/io/minio/MinioClient.java (hunk 2 above)
Comment (balamurugana) [reply api/src/main/java/io/minio/MinioClient.java:2450]: Done

File: functional/FunctionalTest.java (hunk 3 above)
Comment (balamurugana) [reply functional/FunctionalTest.java:2320]: Done

File: api/src/main/java/io/minio/MinioClient.java (hunk 1 above)
Comment (anjalshireesh) [reply api/src/main/java/io/minio/MinioClient.java:2362]:
```suggestion
        builder.metadataDirective(Directive.valueOf(value));
```

**Pull request merged**